src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,240 @@
+        .withArguments("sleep", "hyprlock", "Release fingerprint before sleep", "delay")
+        .storeResultsTo(m_sDBUSState.inhibitLock);
+}
+
+bool CFingerprint::createDeviceProxy() {
+    auto              proxy = sdbus::createProxy(*m_sDBUSState.connection.get(), FPRINT, "/net/reactivated/Fprint/Manager");
+
+    sdbus::ObjectPath path;
+    try {
+        proxy->callMethod("GetDefaultDevice").onInterface(MANAGER).storeResultsTo(path);
+    } catch (sdbus::Error& e) {
+        Debug::log(WARN, "Couldn't connect to Fprint service ({})", e.what());
+        return false;
+    }
+    Debug::log(LOG, "Using dbus path {}", path.c_str());
+    m_sDBUSState.device = sdbus::createProxy(*m_sDBUSState.connection.get(), FPRINT, path);
+
+    m_sDBUSState.device->uponSignal("VerifyFingerSelected").onInterface(DEVICE).call([](const std::string& finger) { Debug::log(LOG, "Finger: {}", finger); });
+    m_sDBUSState.device->uponSignal("VerifyStatus").onInterface(DEVICE).call([this](const std::string& result, const bool done) { handleVerifyStatus(result, done); });
+
+    m_sDBUSState.device->uponSignal("PropertiesChanged")
+        .onInterface("org.freedesktop.DBus.Properties")
+        .call([this](const std::string& interface, const std::map<std::string, sdbus::Variant>& properties) {
+            if (interface != DEVICE || m_sDBUSState.failed)
+                return;
+
+            try {
+                const auto presentVariant = properties.at("finger-present");
+                bool       isPresent      = presentVariant.get<bool>();
+                if (!isPresent)
+                    return;
+                m_sDBUSState.message = m_sFingerprintPresent;
+                g_pHyprlock->enqueueForceUpdateTimers();
+            } catch (std::out_of_range& e) {}
+        });
+
+    m_sDBUSState.device->finishRegistration();
+    return true;
+}
+
+void CFingerprint::handleVerifyStatus(const std::string& result, bool done) {
+    Debug::log(LOG, "Status: {} {}", done, result);
+    auto matchResult = s_mapStringToTestType[result];
+    if (m_sDBUSState.sleeping && matchResult != MatchResult::Disconnected) {
+        return;
+    }
+    switch (matchResult) {
+        case MatchResult::NoMatch:
+            stopVerify();
+            if (m_sDBUSState.retries < 2) {
+                startVerify();
+                done = false;
+                m_sDBUSState.retries++;
+                m_sDBUSState.message = RETRY_MESSAGE;
+            } else {
+                m_sDBUSState.failed  = true;
+                m_sDBUSState.message = "Fingerprint auth disabled: too many failed attempts.";
+            }
+            break;
+        case MatchResult::UnknownError:
+            stopVerify();
+            m_sDBUSState.message = "Unknown fingerprint error.";
+            break;
+        case MatchResult::Match:
+            stopVerify();
+            m_bAuthenticated     = true;
+            m_sDBUSState.message = "";
+            Debug::log(LOG, "matched");
```

Comment:
This log and others could be improved. Ideally prefixed with `fprint: ` or smth.

Reply:
Done